src/ast/binary_encoder.ml

```diff
@@ -0,0 +1,774 @@
+    if is_ref_func then begin
+      write_u32_of_int buf 3;
+      Buffer.add_char buf '\x00';
+      Buffer.add_buffer buf elem_buf
+    end
+    else begin
+      write_u32_of_int buf 7;
+      write_reftype buf ht;
+      Buffer.add_buffer buf elem_buf
+    end
+  | Elem_active (Some 0, expr) ->
+    let elem_buf = Buffer.create 16 in
+    let is_ref_func = write_init elem_buf init in
+    if is_ref_func then write_u32_of_int buf 0 else write_u32_of_int buf 4;
+    write_expr buf expr ~end_op_code:None;
+    Buffer.add_buffer buf elem_buf
+  | Elem_active (Some i, expr) ->
+    let elem_buf = Buffer.create 16 in
+    let is_ref_func = write_init elem_buf init in
+    if is_ref_func then begin
+      write_u32_of_int buf 2;
+      write_indice buf (Raw i);
+      write_expr buf expr ~end_op_code:None;
+      Buffer.add_char buf '\x00';
+      Buffer.add_buffer buf elem_buf
+    end
+    else begin
+      write_u32_of_int buf 6;
+      write_indice buf (Raw i);
+      write_expr buf expr ~end_op_code:None;
+      write_reftype buf ht;
+      Buffer.add_buffer buf elem_buf
+    end
+  | _ -> assert false
+
+let write_data buf ({ init; mode; _ } : data) =
+  match mode with
+  | Data_passive ->
+    write_u32_of_int buf 1;
+    write_string buf init
+  | Data_active (Some 0, expr) ->
+    write_u32_of_int buf 0;
+    write_expr buf expr ~end_op_code:None;
+    write_string buf init
+  | Data_active (Some i, expr) ->
+    write_u32_of_int buf 2;
+    write_u32_of_int buf i;
+    write_expr buf expr ~end_op_code:None;
+    write_string buf init
+  | Data_active (None, _) -> assert false
+
+let encode_section buf id encode_func data =
+  let section_buf = Buffer.create 16 in
+  encode_func section_buf data;
+  let section_len = Buffer.length section_buf in
+  if section_len <> 0 then begin
+    Buffer.add_char buf id;
+    write_u32_of_int buf section_len;
+    Buffer.add_buffer buf section_buf
+  end
+
+(* type: section 1 *)
+let encode_types buf funcs =
+  encode_vector buf funcs
+    (fun buf { type_f = (Bt_raw (_ind, (pt, rt)) : binary block_type); _ } ->
+      Buffer.add_char buf '\x60';
+      write_paramtype buf pt;
+      write_resulttype buf rt )
+
+(* let encode_types_tmp buf { Named.values = types; _ } =
+   encode_vector buf types (fun buf typ ->
+     let typ = Indexed.get typ in
+     match typ with
+     | Def_func_t (pt, rt) ->
+       Buffer.add_char buf '\x60';
+       write_paramtype buf pt;
+       write_resulttype buf rt
+     | _ -> assert false ) *)
+
+(* import: section 2 *)
+let encode_imports buf (funcs, tables, memories, globals) =
+  let imp_buf = Buffer.create 16 in
+  let len =
+    List.length funcs + List.length tables + List.length memories
+    + List.length globals
+  in
+  List.iteri
+    (fun i { Imported.modul; name; _ } ->
+      write_string imp_buf modul;
+      write_string imp_buf name;
+      Buffer.add_char imp_buf '\x00';
+      write_u32_of_int imp_buf i (* TODO ? *) )
```

Comment:
Could you clarify ? :)

Reply:
Refactoring related to new `Binary.modul.types` field : 
See TODO in the new `write_func_import` function